File: backend/test_api.py
import sys
import pytest

# Ensure UTF-8 output on Windows consoles
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")

from fastapi.testclient import TestClient
from main import app

client = TestClient(app)

def test_health_and_info():
    health = client.get("/api/health")
    assert health.status_code == 200
    assert health.json().get("status") == "healthy"

    info = client.get("/api")
    assert info.status_code == 200
    assert "endpoints" in info.json()

def test_trainers():
    trainers = client.get("/api/trainers")
    assert trainers.status_code == 200
    assert len(trainers.json().get("trainers", [])) == 5

    trainer_one = client.get("/api/trainers/tr-1")
    assert trainer_one.status_code == 200
    assert "Vikram" in trainer_one.json().get("trainer", {}).get("name", "")

    slots = client.get("/api/trainers/tr-1/slots")
    assert slots.status_code == 200
    assert len(slots.json().get("schedule", [])) == 7

    trainers_filtered = client.get("/api/trainers?category=Strength")
    assert trainers_filtered.status_code == 200
    assert len(trainers_filtered.json().get("trainers", [])) >= 1

    trainers_search = client.get("/api/trainers?search=Vikram")
    assert trainers_search.status_code == 200
    assert len(trainers_search.json().get("trainers", [])) >= 1

def test_auth():
    reg_payload = {
        "name": "Rohan Sharma",
        "email": "rohan.test@athlete.com",
        "phone": "+91 99999 88888",
        "goal": "Muscle Hypertrophy & Strength"
    }
    register = client.post("/api/auth/register", json=reg_payload)
    assert register.status_code in (200, 201)
    assert register.json().get("user", {}).get("name") == "Rohan Sharma"

    login = client.post("/api/auth/login", json={"email": "rohan.test@athlete.com"})
    assert login.status_code == 200
    assert login.json().get("user", {}).get("email") == "rohan.test@athlete.com"

    profile = client.get("/api/auth/me")
    assert profile.status_code == 200

    update_prof = client.put("/api/auth/profile", json={"phone": "+91 99999 77777"})
    assert update_prof.status_code == 200
    assert update_prof.json().get("user", {}).get("phone") == "+91 99999 77777"

def test_memberships():
    plans = client.get("/api/memberships")
    assert plans.status_code == 200
    assert len(plans.json().get("plans", [])) == 3

    enroll = client.post("/api/memberships/enroll", json={
        "planId": "elite-vip",
        "billingCycle": "annual",
        "paymentMethod": "upi"
    })
    assert enroll.status_code == 200
    assert enroll.json().get("enrollment", {}).get("plan", {}).get("id") == "elite-vip"

    active_tier = client.get("/api/memberships/active")
    assert active_tier.status_code == 200
    assert active_tier.json().get("activeTier") == "elite-vip"

def test_bookings():
    new_booking = client.post("/api/bookings", json={
        "trainerId": "tr-1",
        "date": "2026-09-25",
        "timeSlot": "07:30 AM - 08:30 AM",
        "focus": "Compound Lifting Technique"
    })
    assert new_booking.status_code == 201
    booking_id = new_booking.json().get("booking", {}).get("id")

    bookings_list = client.get("/api/bookings")
    assert bookings_list.status_code == 200
    assert any(b.get("id") == booking_id for b in bookings_list.json().get("bookings", []))

    cancel = client.delete(f"/api/bookings/{booking_id}")
    assert cancel.status_code == 200

def test_facilities_and_inquiries():
    facilities = client.get("/api/facilities")
    assert facilities.status_code == 200
    assert len(facilities.json().get("facilities", [])) > 0

    inquiry = client.post("/api/facilities/inquiry", json={
        "name": "Pooja Verma",
        "email": "pooja@test.com",
        "message": "Interested in annual membership"
    })
    assert inquiry.status_code == 201

def test_reset():
    reset_res = client.post("/api/auth/reset")
    assert reset_res.status_code == 200

if __name__ == "__main__":
    test_health_and_info()
    test_trainers()
    test_auth()
    test_memberships()
    test_bookings()
    test_facilities_and_inquiries()
    test_reset()
    print("All tests passed successfully!")

File: backend/test-api.js
// Automated Smoke Test Suite for PulseFit API Backend
import http from 'http';
import app from './server.js';

process.env.NODE_ENV = 'test';

const TEST_PORT = 5099;
let server;

function request(method, path, body = null) {
  return new Promise((resolve, reject) => {
    const data = body ? JSON.stringify(body) : null;
    const req = http.request(
      {
        hostname: 'localhost',
        port: TEST_PORT,
        path,
        method,
        headers: {
          'Content-Type': 'application/json',
          ...(data && { 'Content-Length': Buffer.byteLength(data) })
        }
      },
      (res) => {
        let raw = '';
        res.on('data', (chunk) => (raw += chunk));
        res.on('end', () => {
          try {
            resolve({
              status: res.statusCode,
              headers: res.headers,
              body: raw ? JSON.parse(raw) : null
            });
          } catch (e) {
            resolve({ status: res.statusCode, headers: res.headers, body: raw });
          }
        });
      }
    );

    req.on('error', reject);
    if (data) req.write(data);
    req.end();
  });
}

let passed = 0;
let failed = 0;

function assert(condition, message) {
  if (condition) {
    console.log(`  ✅ PASS: ${message}`);
    passed++;
  } else {
    console.error(`  ❌ FAIL: ${message}`);
    failed++;
  }
}

async function runTests() {
  console.log('\n🧪 Starting PulseFit API Smoke Tests...\n');

  server = app.listen(TEST_PORT);

  try {
    // 1. Health check
    console.log('Testing System Health & Root Info...');
    const health = await request('GET', '/api/health');
    assert(health.status === 200, 'GET /api/health returned 200');
    assert(health.body.status === 'healthy', 'Health status is "healthy"');

    const info = await request('GET', '/api');
    assert(info.status === 200, 'GET /api returned 200 directory');

    // 2. Trainers
    console.log('\nTesting Trainers Endpoints...');
    const trainers = await request('GET', '/api/trainers');
    assert(trainers.status === 200, 'GET /api/trainers returned 200');
    assert(trainers.body.trainers.length === 5, 'Returns 5 master coaches');

    const trainerOne = await request('GET', '/api/trainers/tr-1');
    assert(trainerOne.status === 200, 'GET /api/trainers/tr-1 returns coach details');
    assert(trainerOne.body.trainer.name.includes('Vikram'), 'Coach name is Vikram "Titan" Rathore');

    const slots = await request('GET', '/api/trainers/tr-1/slots');
    assert(slots.status === 200, 'GET /api/trainers/tr-1/slots returns 7-day schedule');
    assert(slots.body.schedule.length === 7, 'Schedule has 7 calendar days');

    // 3. Auth
    console.log('\nTesting Athlete Auth & Profile...');
    const register = await request('POST', '/api/auth/register', {
      name: 'Rohan Sharma',
      email: 'rohan.test@athlete.com',
      phone: '+91 99999 88888',
      goal: 'Muscle Hypertrophy & Strength'
    });
    assert(register.status === 201 || register.status === 200, 'POST /api/auth/register creates user');
    assert(register.body.user.name === 'Rohan Sharma', 'User name matches registered input');

    const login = await request('POST', '/api/auth/login', {
      email: 'rohan.test@athlete.com'
    });
    assert(login.status === 200, 'POST /api/auth/login returns user session');

    const profile = await request('GET', '/api/auth/me');
    assert(profile.status === 200, 'GET /api/auth/me returns active session');

    // 4. Memberships
    console.log('\nTesting Membership Tiers...');
    const plans = await request('GET', '/api/memberships');
    assert(plans.status === 200, 'GET /api/memberships returns plans');
    assert(plans.body.plans.length === 3, 'Returns Starter, Pro Beast, and Elite VIP tiers');

    const enroll = await request('POST', '/api/memberships/enroll', {
      planId: 'elite-vip',
      billingCycle: 'annual',
      paymentMethod: 'upi'
    });
    assert(enroll.status === 200, 'POST /api/memberships/enroll activates plan');
    assert(enroll.body.enrollment.plan.id === 'elite-vip', 'Enrolled in Elite VIP tier');

    // 5. Bookings
    console.log('\nTesting Trainer Slot Bookings...');
    const newBooking = await request('POST', '/api/bookings', {
      trainerId: 'tr-1',
      date: '2026-09-25',
      timeSlot: '07:30 AM - 08:30 AM',
      focus: 'Compound Lifting Technique'
    });
    assert(newBooking.status === 201, 'POST /api/bookings reserves 60-min slot');
    const bookingId = newBooking.body.booking.id;

    const bookingsList = await request('GET', '/api/bookings');
    assert(bookingsList.status === 200, 'GET /api/bookings returns booked list');
    assert(bookingsList.body.bookings.some((b) => b.id === bookingId), 'Newly booked slot exists in list');

    const cancel = await request('DELETE', `/api/bookings/${bookingId}`);
    assert(cancel.status === 200, 'DELETE /api/bookings/:id releases booked slot');

    // 6. Facilities & Inquiries
    console.log('\nTesting Facilities & Inquiries...');
    const facilities = await request('GET', '/api/facilities');
    assert(facilities.status === 200, 'GET /api/facilities returns facility details');

    const inquiry = await request('POST', '/api/facilities/inquiry', {
      name: 'Pooja Verma',
      email: 'pooja@test.com',
      message: 'Interested in annual membership'
    });
    assert(inquiry.status === 201, 'POST /api/facilities/inquiry records contact inquiry');

    console.log(`\n========================================`);
    console.log(`🎯 Test Summary: ${passed} Passed, ${failed} Failed`);
    console.log(`========================================\n`);

    if (failed > 0) {
      process.exitCode = 1;
    }
  } catch (err) {
    console.error('Smoke tests crashed with error:', err);
    process.exitCode = 1;
  } finally {
    server.close();
  }
}

runTests();
